pkg/edgeview/src/copyfile.go

```diff
@@ -0,0 +1,359 @@
+// Copyright (c) 2021 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package main
+
+import (
+	"archive/tar"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+const (
+	startCopyMessage  = "+++Start-Copy+++"
+	fileCopyDir       = "/download/"
+)
+
+var (
+	isCopy        bool           // client side
+	isSvrCopy     bool           // server side
+	copyMsgChn    chan []byte
+	lastPerc      int64
+	barDone       chan struct{}
+)
+
+type copyFile struct {
+	TokenHash []byte   `json:"tokenHash"`
+	Name      string   `json:"name"`
+	Size      int64    `json:"size"`
+	Sha256    string   `json:"sha256"`
+	ModTsec   int64    `json:"modtsec"`
+}
+
+type fileCopyStatus struct {
+	gotFileInfo   bool
+	filename      string
+	fileSize      int64
+	fileHash      string
+	currSize      int64
+	modTime       time.Time
+	buf           []byte
+	f             *os.File
+}
+
+func runCopy(opt string) {
+	path := strings.SplitN(opt, "cp/", 2)
+	if len(path) != 2 {
+		fmt.Printf("cp needs a cp/ and path input\n")
+		return
+	}
+	file := path[1]
+	info, err := os.Stat(file)
```

Comment:
This SplitN is OK since it implicitly verifies that it is a valid filename. But I don't know if one can make that more explicit by calling filepath.Abs() on the name. From what I can tell that still doesn't check that it is a syntactically valid pathname so maybe just using Stat() is sufficient.

FWIW I'm looking at all the SplitN() calls in this PR.